### PR TITLE
Remove self references in CommonServerPowershell 

### PR DESCRIPTION
**Pull request by DinaMeylakh**
## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: [link to the issue](https://jira-dc.paloaltonetworks.com/browse/XSUP-30551)

## Description
Remove self referencing properties from raw results. They are recursively converted to json causing duplication in results.

## Must have
- [x] Tests
- [ ] Documentation

### CODE REVIEW

File: Packs/Base/Scripts/CommonServerPowerShell/CommonServerPowerShell.ps1 [hunk 1]
```diff
@@ -401,7 +482,7 @@ function ReturnOutputs([string]$ReadableOutput, [object]$Outputs, [object]$RawRe
         Type           = [EntryTypes]::note;
         ContentsFormat = [EntryFormats]::json.ToString();
         HumanReadable  = $ReadableOutput;
-        Contents       = $RawResponse;
+        Contents       = Remove-SelfReferences $RawResponse;
         EntryContext   = $Outputs
```
Comment (DinaMeylakh) [inline Packs/Base/Scripts/CommonServerPowerShell/CommonServerPowerShell.ps1:486]: Add flag to Remove-Self References in this function and add it to outputs as well (if flag is true).

File: Packs/Base/Scripts/CommonServerPowerShell/CommonServerPowerShell.ps1 [hunk 2]
```diff
@@ -35,6 +35,87 @@ enum EntryFormats {
...
 }
 
+function Get-SelfReferencingPaths($obj, $visited = @{}, $path = @(), $depth = 1, $maxDepth = 20) {
+    if ($depth -gt $maxDepth) {
+        # Stop recursion when max depth is reached
+        return @()
+    }
+
+    $selfReferencingPaths = @()
+    if ($obj | Get-Member -MemberType Properties -ErrorAction SilentlyContinue) {
```
Comment (DinaMeylakh) [inline Packs/Base/Scripts/CommonServerPowerShell/CommonServerPowerShell.ps1:45]: add comment explaining this if

File: Packs/Base/Scripts/CommonServerPowerShell/CommonServerPowerShell.ps1 [hunk 3]
```diff
@@ -35,6 +35,87 @@ enum EntryFormats {
...
+        $visited[$obj] = $true
+
+        foreach ($property in $obj.PSObject.Properties) {
+            $propertyValue = $property.Value
+            $propertyPath = "$($path).$($property.Name)"
+            if (-not $path) {
+                $propertyPath = "$($property.Name)"
+            }
+
+            if (-not ($propertyValue -is [String] -or $propertyValue -is [Int32] -or $propertyValue -is [Double] -or $propertyValue -is [DateTime] -or $propertyValue -eq $null)) {
```
Comment (DinaMeylakh) [inline Packs/Base/Scripts/CommonServerPowerShell/CommonServerPowerShell.ps1:65]: Add comment explaining this If

Comment (DinaMeylakh) [reply Packs/Base/Scripts/CommonServerPowerShell/CommonServerPowerShell.ps1:65]: move/merge this if with the one above

File: Packs/Base/Scripts/CommonServerPowerShell/CommonServerPowerShell.Tests.ps1 [hunk 4]
```diff
@@ -332,4 +332,106 @@ Describe "Test ParseDateRange"{
         It "Wrong time unit" {
             { ParseDateRange("3 planets") } | Should -Throw
         }
-    }
\ No newline at end of file
+    }
+
+Describe "Test Remove-SelfReferences" {
```
Comment (DinaMeylakh) [inline Packs/Base/Scripts/CommonServerPowerShell/CommonServerPowerShell.Tests.ps1:337]: add documentation to tests

**Review by ShahafBenYakir — APPROVED** [review]
Great

**Pull request merged**